crypto/x509/x509_vfy.c

```diff
@@ -1756,6 +1756,10 @@ static int internal_verify(X509_STORE_CTX *ctx)
         goto check_cert_time;
     }
 
+    /* Make sure that the root of the chain is trusted */
+    CHECK_CB(!X509_check_trust(xi, ctx->param->trust, ctx->param->flags),
+             ctx, xi, n, X509_V_ERR_CERT_UNTRUSTED);
```

Comment:
I think this check is misleading.  The certificate in question is now detached from its provenance, we don't know whether it came from the trust store, or from the peer-supplied untrusted chain!

That would not be the case if trust anchors always had trust EKUs (perhaps as part of import into the store?), and otherwise, certificates from the untrusted chain were sure to not have them (guaranteed with SSL I believe, but not necessarily when read from PEM files, depending on how they're read).

So I am not sure this is a good idea...

Reply:
Thanks for your comment. 
I was actually mislead by the name of the (BTW public but undocumented) function `X509_check_trust()`, 
although I should have noticed that it does not take into account the trust store!
So the check I introduced was too weak. 

I've fixed this now, and on this occasion improved the behavior of `lookup_cert_match` w.r.t `up_ref()`.

Reply:
With the stronger check, now one of the DANE tests fails:
```
===(    2641;296  25/90   4/15   1/11 )=================================
    # ERROR: (int) 'err == want' failed @ test/danetest.c:363
    # [27] compared to [0]
    # INFO: 
    # Verification failure in test 13: 27=certificate not trusted
    # ERROR: (int) 'test_tlsafile(ctx, basedomain, f, tlsafile) > 0' failed @ test/danetest.c:401
    # [0] compared to [0]
    not ok 1 - run_tlsatest
# ------------------------------------------------------------------------------
../../util/wrap.pl ../../test/danetest example.com ../../test/danetest.pem ../../test/danetest.in => 1
not ok 1 - dane tests
===(    2642;296  25/90   4/15   1/11  1/1 )============================--------
#   Failed test 'dane tests'
#   at test/recipes/80-test_dane.t line 22.
80-test_dane.t ..................... Dubious, test returned 1 (wstat 256, 0x100)
```
which I find surprising because the DANE bare TA case has already been handled before in `internal_verify()`:
```
    /*
     * With DANE-verified bare public key TA signatures, it remains only to
     * check the timestamps of the top certificate.  We report the issuer as
     * NULL, since all we have is a bare key.
     */
    if (ctx->bare_ta_signed) {
        xs = xi;
        xi = NULL;
        goto check_cert_time;
    }

```

Reply:
> I've fixed this now by adding `lookup_cert_match()` to make sure that (regardless where the root cert originated from) it directly matches a cert in the trust store and is trusted.
> On this occasion improved the behavior of `lookup_cert_match()` regarding `up_ref()`.

DANE-TA(2) explicitly bypasses the trust store, a chain is valid iff the peer-provided certificates result in a chain where one of the issuer CAs (non-leaf-certs) matches a "TLSA 2 X X" record (it or the EE cert) is signed directly by a bare public key from a "2 1 0" record.  So such a check does not belong in internal_verify() (whose job is to check the signature chain and expiration dates, i.e. just chain validity, and NOT trust).

Reply:
Thanks for this explanation. So DANE has a more flexible way of putting trust, namely potentially relating to any of the members of the chain. I suppose `internal_verify()` does not (re-)check this trust; is there a function that could be used to directly (re-)check the trust for the DANE case independently of chain building?

For the non-DANE case I've move the proposed check for direct trust in the root of the chain out of `internal_verify()` to `verify_chain()` and made sure it is not called for DANE.
This makes all tests pass again.